src/liballoc/borrow.rs

```diff
@@ -479,3 +479,18 @@ impl<'a> AddAssign<Cow<'a, str>> for Cow<'a, str> {
         }
     }
 }
+
+impl<'a> AddAssign<char> for Cow<'a, str> {
+    fn add_assign(&mut self, rhs: char) {
+        if self.is_empty() {
+            *self = rhs.to_string()
```

Comment:
Why treat the empty string case specially? Remember that `rhs` is a `char`, so we're not reusing an existing string.

We have three cases:

1. Empty borrowed `str`: the `Cow::Borrowed` code below does the right thing by allocating a new string with the correct capacity.

2. Owned `String` with zero-capacity: the push allocates appropriately.

3. Owned `String` with non-zero capacity: by throwing it away we're wasting the allocation, and surprising the user who would have expected that pre-allocated capacity to be used. Better to not treat this as a special case, and just do a push as usual.

Reply:
@petertodd Thank you for pointing this out. We indeed do not need to have new allocations in such cases.